New rule: `no-meta-replaced-by` (#518)
eslint-community/eslint-plugin-eslint-plugin: docs/rules/no-meta-replaced-by.md

Reviewer: > Would it be appropriate to write
> 
> ```
> ## When Not To Use It
> 
> If you need to support versions of ESLint prior to v9.
> ```

Regarding when not to use it, I don't think we need to include this section. In fact, anyone can use this rule regardless of ESLint version. As far as I know, ESLint itself does not validate nor use the `meta.deprecated` property. So anybody can use the object format for `meta.deprecated`.

Note that if someone is using TypeScript to define their rules, they would want to make sure they have up-to-date rule types (which I believe come from @types/eslint or typescript-eslint). And rule authors would want to make sure they are using up-to-date versions of any third-party tooling like [eslint-doc-generator](https://github.com/bmish/eslint-doc-generator) that use this property.

# Disallow rules to use a `meta.replacedBy` property (`eslint-plugin/no-meta-replaced-by`)

💼 This rule is enabled in the ✅ `recommended` [config](https://github.com/eslint-community/eslint-plugin-eslint-plugin#presets).

<!-- end auto-generated rule header -->

> ESLint v9.21.0 introduces a new `DeprecatedInfo` type to describe whether a rule is deprecated and how it can be replaced. The legacy format used the meta property `replacedBy` which should be defined inside `deprecated` instead of at the top level.

Reviewer: ```suggestion
As of ESLint v9.21.0, the rule property `meta.deprecated` can be either a boolean or an object of type `DeprecatedInfo`. The `DeprecatedInfo` includes an optional `replacedBy` array that replaces the now-deprecated `meta.replacedBy` property.
```

> Examples of a correct usage can be found at [array-bracket-newline](https://github.com/eslint/eslint/blob/4112fd09531092e9651e9981205bcd603dc56acf/lib/rules/array-bracket-newline.js#L18-L38) and [typescript-eslint/no-empty-interface](https://github.com/typescript-eslint/typescript-eslint/blob/af94f163a1d6447a84c5571fff5e38e4c700edb9/packages/eslint-plugin/src/rules/no-empty-interface.ts#L19-L30)

Reviewer: ```suggestion
Examples of correct usage:

- [array-bracket-newline](https://github.com/eslint/eslint/blob/4112fd09531092e9651e9981205bcd603dc56acf/lib/rules/array-bracket-newline.js#L18-L38)
- [typescript-eslint/no-empty-interface](https://github.com/typescript-eslint/typescript-eslint/blob/af94f163a1d6447a84c5571fff5e38e4c700edb9/packages/eslint-plugin/src/rules/no-empty-interface.ts#L19-L30)
```

## Rule Details

> This rule disallows the `replacedBy` property in rules' `meta`.

Reviewer: ```suggestion
This rule disallows the `meta.replacedBy` property in a rule.
```

Examples of **incorrect** code for this rule:

```js
/* eslint eslint-plugin/no-meta-replaced-by: error */

module.exports = {
  meta: {
    deprecated: true,

>     replacedBy: [],

Reviewer: ```suggestion
    replacedBy: ['the-new-rule'],
```

  },
  create() {},
};
```

Examples of **correct** code for this rule:

```js
/* eslint eslint-plugin/no-meta-replaced-by: error */

module.exports = {
  meta: {
    deprecated: {
      message: 'The new rule adds more functionality',
      replacedBy: [
        {
          rule: {
            name: 'the-new-rule',
          },
        },
      ],
    },
  },
  create() {},
};

module.exports = {
  meta: {},
  create() {},
};
```

## When Not To Use It

If you do not plan to provide rule's documentation in website, you can turn off this rule.

## Further Reading

- [ESLint docs: `DeprecatedInfo`](https://eslint.org/docs/latest/extend/rule-deprecation#-deprecatedinfo-type)

> - [RFC introducing `DeprecatedInfo` type](https://github.com/eslint/rfcs/blob/main/designs/2023-rule-options-defaults/README.md)

Reviewer: Wrong link

# Disallow using the `meta.replacedBy` rule property (`eslint-plugin/no-meta-replaced-by`)

<!-- end auto-generated rule header -->

> As of ESLint v9.21.0, the rule property `meta.deprecated` can be either a boolean or an object of type `DeprecatedInfo`. The `DeprecatedInfo` includes an optional `replacedBy` array that replaces the now-deprecated `meta.replacedBy` property.

Reviewer: ```suggestion
As of ESLint v9.21.0, the rule property `meta.deprecated` can be either a boolean or an object of type `DeprecatedInfo`. The `DeprecatedInfo` type includes an optional `replacedBy` array that replaces the now-deprecated `meta.replacedBy` property.
```
```suggestion
As of ESLint v9.21.0, the rule property `meta.deprecated` can be either a boolean or an object of type `DeprecatedInfo`. The `DeprecatedInfo` includes an optional `replacedBy` array that replaces the now-deprecated `meta.replacedBy` property.
```

Author: Err ... github display two suggested changes. One with and another without `type` in the second sentence (?)

Reviewer: fixed, want to add `type`.

Examples of correct usage:

- [array-bracket-newline](https://github.com/eslint/eslint/blob/4112fd09531092e9651e9981205bcd603dc56acf/lib/rules/array-bracket-newline.js#L18-L38)
- [typescript-eslint/no-empty-interface](https://github.com/typescript-eslint/typescript-eslint/blob/af94f163a1d6447a84c5571fff5e38e4c700edb9/packages/eslint-plugin/src/rules/no-empty-interface.ts#L19-L30)

## Rule Details

This rule disallows the `meta.replacedBy` property in a rule.

Examples of **incorrect** code for this rule:

```js
/* eslint eslint-plugin/no-meta-replaced-by: error */

module.exports = {
  meta: {
    deprecated: true,
    replacedBy: ['the-new-rule'],
  },
  create() {},
};
```

Examples of **correct** code for this rule:

```js
/* eslint eslint-plugin/no-meta-replaced-by: error */

module.exports = {
  meta: {
    deprecated: {
      message: 'The new rule adds more functionality',
      replacedBy: [
        {
          rule: {
            name: 'the-new-rule',
          },
        },
      ],
    },
  },
  create() {},
};

module.exports = {
  meta: {},
  create() {},
};
```

> ## When Not To Use It
> 
> If you do not plan to provide rule's documentation in website, you can turn off this rule.
> 

Reviewer: ```suggestion
```
